sklearn/metrics/pairwise.py

```diff
@@ -1275,6 +1440,13 @@ def _check_chunk_size(reduced, chunk_size):
 def _precompute_metric_params(X, Y, metric=None, **kwds):
     """Precompute data-derived metric parameters if not provided
     """
+    if metric == "gower":
+        MIN_MAX = {}
+        if 'MIN' not in kwds:
+            MIN_MAX['MIN'] = np.nanmin((np.vstack([X, Y]), X)[X is Y], axis=0)
```

Comment:
I'm not certain, but I think we should always just get the scale from X and ignore Y (and document and test that behaviour).

You are wasting memory here on the stacking, also.

Reply:
@jnothman, @adrinjalali, I've been implementing the support to range in X and Y so far, and there are documentation and tests supporting this:
https://github.com/scikit-learn/scikit-learn/pull/9555#discussion_r154647818
https://github.com/scikit-learn/scikit-learn/pull/9555#discussion_r181001223

If I remove the range support of Y, we'll get incorrect values of Gower, if Y is not comprised in the range of X, scale=true, and there is no range informed by the user. Once removed I won't put this Y range code back in this PR.

Please decide if you really want to remove the range support of Y, then I remove everything, but I guess we are just creating conditions for a bug in the future.

Reply:
I'm not sure what it means to get "incorrect values of Gower". Do you mean that some fundamental property of the metric will be violated?

I don't think the metric was designed with concern for "what happens with new data points?"

My problem with the current solution is that `gower_distances(X, Y[:1])` may give results inconsistent with `gower_distances(X, Y[:2])`. In a train-test paradigm, this seems a big issue (if we assume that X corresponds to training data, which it does in our brute force nearest neighbours implementation unless I'm much mistaken).

Reply:
However: the current solution is consistent with what is already done in _precompute_metric_params for other metrics, so we should keep it that way and I should raise a new issue about the handling of Y in _precompute_metric_params

Reply:
>> My problem with the current solution is that gower_distances(X, Y[:1]) may give results inconsistent with gower_distances(X, Y[:2]). 
I understand that Y[:1] and Y[:2] are different datasets, in the proposed implementation the user can pass the ranges of X, Y anytime and the problem is solved, removing the automatic scaling of Y won't solve this problem in any case, I think one problem is not related to another.

>> Do you mean that some fundamental property of the metric will be violated?
Exactly, you will get values not between 0 and 1 as result, if the values of Y is not between X, and this violates the proposal of Gower formula (very top of page 860 in the original paper).

I'm just proposing to make an automatic scaling of Y here, the user can push the ranges they want in the scale parameter anytime. 

It's up to the reviewers if we are going to keep the automatic scaling of Y. 
Please confirm if we must remove the implementation, documentation, and tests for automatic scaling of Y.

Reply:
> I understand that Y[:1] and Y[:2] are different datasets

I don't understand your comment here. `Y[0]` is the same sample in both those datasets. In your implementation `gower_distances(X, Y[:1])[:, 0]` may not be equal to `gower_distances(X, Y[:2])[:, 0]` for fixed `X` and `Y`. That is bad behaviour to me.

> Exactly, you will get values not between 0 and 1 as result,

Bounding a distance at 1 is a fairly unusual feature of Gower. I'd rather have some distances exceed 1 and have the awkwardness visible than invisibly have some distances change depending on the other samples in the test dataset.

But we'll keep using X and Y because this is what `_precompute_metric_params` does (problematically IMO).

Reply:
>> In your implementation gower_distances(X, Y[:1])[:, 0] may not be equal to gower_distances(X, Y[:2])[:, 0] for fixed X and Y. 

The same can happen to gower_distances(X[:1])[:, 0] that may not be equal to gower_distances(X[:2])[:, 0] for fixed X only. As I said, the problem here is not scaling X, Y. Thankfully you proposed to have an optional scale, or pass the ranges.

Reply:
> The same can happen to gower_distances(X[:1])[:, 0] that may not be equal to gower_distances(X[:2])[:, 0] for fixed X only. 

But if we, as the masters of our API, were to declare that X corresponds to the training/reference data, and Y to new samples, or vice versa, then that wouldn't be a problem. It looks like Nearest Neighbors actually uses Y as the reference.

Let's drop this issue for now. I'll raise it elsewhere; I just haven't got around to it yet.